pyquil/paulis.py

```diff
@@ -407,6 +414,74 @@ def pauli_string(self, qubits=None):
 
         return ''.join(self[q] for q in qubits)
 
+    def matrix(self, qubit_mapping={}, nqubits=None, check_mapping=True):
+        """Create the matrix representation of self.
+
+        :param nqubits:  Number of qubits to represent self on. If none the maximum qubit in the
+                         PauliTerm or qubit_mapping is used.
+        :param qubit_mapping: A dictionary-like object that maps from :py:class`QubitPlaceholder`
+                              to :py:class:`int`
+        :check_mapping: (bool) whether or not to check qubit_mapping for validity
```

Comment:
As far as I see it, it does use the qubit_mapping in line 441 and line 442.
But I was also wondering, if we want to go through all the effort to check validity even in the case, where the PauliTerm has real qubits and QubitPlaceholders. In the code nothing stops me from creating such mixed PauliTerms. But in the case of Program() the latest at the time I try to compile it, the compiler will complain about mixing instantiated Qubits Qubitplaceholders. So maybe just assume, that PauliTerm will always have only QubitPlaceholders or instantiated Qubits?

Reply:
"As far as I see it, it does use the qubit_mapping in line 441 and line 442." Correct, my bad. Disregard comment.

Reply:
Your next point is a good one. If I've understood it correctly, you're asking if there should be some sort of error raising if we use both QubitPlaceholders and ints? We also don't get an error msg if we do something similar for Programs, so e.g. if I do
```
from pyquil.quil import address_qubits
from pyquil.quilatom import QubitPlaceholder
from pyquil.gates import H, CNOT
from pyquil import Program


q0 = QubitPlaceholder()
prog = Program(H(q0), CNOT(q0, 1))
```
I won't get any complaints from pyquil. I'll only get an error msg when I try to do
```
prog_addressed = address_qubits(prog, qubit_mapping={q0: 0})
```
which gives me a `ValueError: Your program mixes instantiated qubits with placeholders`.

So I think it makes sense to be able to hold both QubitPlaceholders and real qubits in a PauliTerm, especially since `address_qubits` doesn't convert QubitPlaceholders in PauliTerms to real qubits (so you might have to write a custom mapping yourself).

Perhaps others might have thoughts about this too. @notmgsk @karalekas @ecpeterson ?